File: src/flydocs/interfaces/dtos/extraction.py
# Copyright 2024-2026 Firefly Software Foundation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""DTOs for the async extraction lifecycle.

Endpoints:

* ``POST   /api/v1/extractions``              -- submit
* ``GET    /api/v1/extractions``              -- list
* ``GET    /api/v1/extractions/{id}``         -- status
* ``GET    /api/v1/extractions/{id}/result``  -- final result envelope
* ``DELETE /api/v1/extractions/{id}``         -- cancel (only while queued)

The main lifecycle is linear: ``queued -> running -> succeeded | failed |
cancelled``. Post-processing (bbox refinement today, more tomorrow) lives in
the additive :class:`PostProcessing` block with its own
:class:`PostProcessingStatus` lifecycle that does not gate the main status.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import AnyHttpUrl, BaseModel, ConfigDict, Field

from flydocs.interfaces.dtos.extract import ExtractionRequest, ExtractionResult
from flydocs.interfaces.enums.extraction_status import ExtractionStatus, PostProcessingStatus


class SubmitExtractionRequest(ExtractionRequest):
    """Submit shape: full extraction request plus async-only fields."""

    callback_url: AnyHttpUrl | None = Field(
        default=None,
        description="If set, the worker POSTs an EventEnvelope here on terminal status.",
    )
    metadata: dict[str, Any] = Field(default_factory=dict)


class ExtractionError(BaseModel):
    """Terminal-state error info for a failed extraction."""

    model_config = ConfigDict(extra="forbid")

    code: str
    message: str


class BboxRefinementInfo(BaseModel):
    """Lifecycle info for the bbox-refinement post-processing leg."""

    model_config = ConfigDict(extra="forbid")

    status: PostProcessingStatus
    started_at: datetime | None = None
    finished_at: datetime | None = None
    attempts: int = 0
    error: ExtractionError | None = None


class PostProcessing(BaseModel):
    """Container for post-processing legs attached to a succeeded extraction."""

    model_config = ConfigDict(extra="forbid")

    bbox_refinement: BboxRefinementInfo | None = None


class Extraction(BaseModel):
    """Current state snapshot of an async extraction job."""

    model_config = ConfigDict(extra="forbid")

    id: str
    status: ExtractionStatus
    submitted_at: datetime
    started_at: datetime | None = None
    finished_at: datetime | None = None
    attempts: int = 0
    error: ExtractionError | None = None
    post_processing: PostProcessing | None = None


class ExtractionResultEnvelope(BaseModel):
    """``GET /extractions/{id}/result`` body."""

    model_config = ConfigDict(extra="forbid")

    id: str
    result: ExtractionResult


class ExtractionListQuery(BaseModel):
    """Query parameters for ``GET /api/v1/extractions``."""

    model_config = ConfigDict(extra="forbid")

    statuses: list[ExtractionStatus] = Field(default_factory=list)
    post_processing_statuses: list[PostProcessingStatus] = Field(default_factory=list)
    created_after: datetime | None = None
    created_before: datetime | None = None
    idempotency_key: str | None = None
    limit: int = Field(default=50, ge=1, le=500)
    offset: int = Field(default=0, ge=0)


class ExtractionListResponse(BaseModel):
    """Paginated list response."""

    model_config = ConfigDict(extra="forbid")

    items: list[Extraction]
    total: int
    limit: int
    offset: int

File: sdks/java/flydocs-sdk/src/main/java/com/firefly/flydocs/sdk/model/Extraction.java
// Copyright 2024-2026 Firefly Software Foundation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.firefly.flydocs.sdk.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.time.OffsetDateTime;
import org.jspecify.annotations.Nullable;

/**
 * Current state snapshot of an async extraction.
 *
 * <p>Response body for {@code POST /api/v1/extractions} (202),
 * {@code GET  /api/v1/extractions/{id}}, and {@code DELETE  /api/v1/extractions/{id}}.</p>
 *
 * <p>v1 simplifies the state machine: {@code queued -> running -> succeeded
 * | failed | cancelled}. The v0 {@code PARTIAL_SUCCEEDED} / {@code REFINING_BBOXES}
 * intermediate states are gone; bbox refinement runs as additive
 * post-processing under {@link #postProcessing()} on a fully-succeeded
 * extraction.</p>
 */
@JsonIgnoreProperties(ignoreUnknown = true)
@JsonInclude(JsonInclude.Include.NON_NULL)
public record Extraction(
        @JsonProperty("id") String id,
        @JsonProperty("status") ExtractionStatus status,
        @JsonProperty("submitted_at") OffsetDateTime submittedAt,
        @JsonProperty("started_at") @Nullable OffsetDateTime startedAt,
        @JsonProperty("finished_at") @Nullable OffsetDateTime finishedAt,
        @JsonProperty("attempts") int attempts,
        @JsonProperty("error") @Nullable ExtractionError error,
        @JsonProperty("post_processing") @Nullable PostProcessing postProcessing) {

    /** True when no further main-status transition is expected. */
    public boolean isTerminal() {
        return status != null && status.isTerminal();
    }
}
